### PR TITLE
feat: refactor FSAL to separate file tree and abstraction logic

### LINKED ISSUE #4337: Bugs in the FSAL updating process
### Description

Sometimes – especially when modifying many different files via the Explorer/Finder/file browser (i.e., remotely from the perspective of Zettlr), the file tree that is displayed in the file manager is no longer in sync with what is actually on disk.

### Reproducing

Zettlr is running. Then you create a new folder in any of the loaded workspaces in Zettlr and in it put a dozen different files, and afterward you rename those files, one by one. Then, when you go back to the Zettlr file manager to check what it displays, you notice that it hasn't updated everything. So what Zettlr displays you is not necessarily what is actually on disk.

### Zettlr Version

Right off the develop branch

### Specify version

v3.0.0-beta.4-develop

### Installation Method

None

### Your Platform

- [ ] Windows
- [X] macOS
- [ ] Linux

### Architecture

- [ ] Intel 64bit
- [X] ARM 64bit

### Operating System Version

macOS Ventura 13.2.1

### Additional Information

I believe this bug to be partially related to how we manage updates in the renderer process. Currently, it looks like this:

1. Something changes remotely
2. The FSAL notices this and updates its own file tree
3. It records events of everything added, changed, or deleted
4. It sends a notification to the renderer process that something has changed
5. The renderer then requests all events (additions, changes, deletions) that happened since the last time it asked and attempts to apply them one by one

I believe that in step 5, sometimes there are still problems since sometimes files get added and then immediately deleted, which make it difficult for the logic to keep up.

I believe that a better logic should be found that isn't as brittle as the current one. Specifically, I am currently pondering of whether the following would be sufficient (especially in terms of speed):

1. Something changes remotely
2. The FSAL notices this and updates its own file tree
3. It sends a notification to the renderer process of the minimal subtree it has to reload to be back in sync
4. The renderer requests this minimal subtree and exchanges it in its own representation
5. Should that fail, the renderer attempts a full recovery of the workspace in question

This would ensure that the renderer is not always requesting the entire tree (which can be Megabytes in size) but rather the minimum necessity without being too granular. A subtree should only in worst cases take noticeable time to get changed, a workspace will lead to a small delay, but I think that would be worth it, especially given that the FSAL has always had these synchronization problems.

### PR DESCRIPTION
**Pull request by nathanlesage**
This PR is now ready for review.

I have also tagged you, @sensologica, in one place where you can, if you want to, start thinking about nice ways of implementing that (search for your name below).

## Description

This PR fulfills a somewhat older dream of mine: The actual creation of a _File System Abstraction Layer_.

Some background: When I wrote the FSAL, I intended to create a class that would abstract any file access away from the application logic, in a somewhat similar way as Laravel's (or Symfony's?) storage providers work.

The main idea was simple: Centralize all file system logic in one place, and locate the business logic someplace else. This would also give me the chance to easily support remote file systems, such as WebDav, Dropbox, Google Drive, or One Drive (if I at some point in my life find the time to do that lol).

Long story short: I wanted to call `loadFile(filePath)` somewhere in my app, and the `fs.readFile`-stuff would happen in the FSAL.

Now, that was my goal. Which I missed. By about a mile. The thing is, back when I implemented the FSAL, the app was not quite ready, since I had two problems:

1. fsevents didn't work back then ([read here](https://www.hendrik-erz.de/post/electron-chokidar-and-native-nodejs-modules-a-horror-story-from-integration-hell) on the entire sh\*\*hole that was the fsevents integration).
2. I had no place where I would collect and manage the file tree.

Problem (1) led to a ton of arbitrary logic of patching the file tree in numerous creative ways. The issue was that I didn't trust fsevents (and myself) back then. This led to all the `ignoreEvents` shenanigans that have, in my humble opinion, led to bloated code in the FSAL, because especially the `move` action required a ton of additional code and absurd amounts of events being broadcasted across the entire app.

Problem (2) led to me actually violating the FSAL name: Instead of just abstracting away the file system logic, I also stuffed the entire file tree management logic in there. Which led to even more bloat.

In effect, these two problems I faced back in the day led to a class that is currently well above 1,000 lines of code and that has tons of problems. These problems mainly stem from the insufficient edge case handling in the brittle patching logic that I had thus far. Users oftentimes complain (and rightly so) that some events just don't get processed by the app and in the worst case, users would need to restart the app.

I had the honor once when I checked out an older commit in a loaded git repository. Git's moving back and forth of tons of files has just defeated the state management in the under-equipped FSAL.

With this PR, I hope to address most of these issues. It is on the one hand a refactor, and on the other hand an entire reconceptualization of file management in the app.

## Changes

On a high level, this PR changes only two things:

1. Remove the entire logic of file tree management out of the FSAL and into their own classes (`Root`). This entirely separates FS management and file tree management logic.
2. Move the entire responsibility of actually detecting changes to the chokidar library, and not tamper with anything it does.

In addition, this PR has a series of very neat side effects:

1. This PR separates all loaded roots from each other. This will make implementing a very nice feature easy: Enable users to use windows as a form of "Workspaces" similar to VSCode, where they can choose which folders they see, and which not. I myself figured out that I really want to visually separate my reading notes and Zettelkasten directories (where I am not interested in browsing files inasmuch as I want to search them) from my PhD directory (where everything follows a set folder structure), so that's now a possibility. **Possibly interesting for @sensologica.**
2. This logic is inspired a lot from CodeMirror's new collab plugin and makes it easy to manage file trees in any combination
3. **This PR begins the switch to Pinia as already suggested years ago by @tobiasdiez.** The new logic has been entirely implemented in Pinia stores which, for now, live side by side with the Vuex store. This makes the entire logic in the renderer type safe and enables us to at some point get rid of the very last JavaScript-file in the project (`item-mixin.js`).
4. This PR now centralizes all file system logic. Specifically, I took the liberty of going across the application and redirecting any arbitrary file access from direct access of the  `fs` module to the FSAL. There are a few places where `isDir` and `isFile` still need to be called, but those are comparatively safe now. This reduces the amount of synchronous file access to almost zero across the runtime of the app.

## Additional information

I just want to highlight how insanely simple implementation was. This goes to show how simple programming can be if your data model is appropriate and fit to the task. For years, I was grappling with implementing billions of state changes, but implementing the queue for file tree changes was a breeze. Thanks at this point to the anonymous redditor, who recommended "Threadsafe queues" to me, which I didn't end up implementing but which gave me the final inspiration to pull through with this PR.

* Closes #4337

Tested on: macOS Sonoma 14.1.1 (MBP 14'' M2 Pro)

## Known Bugs

I have now merged the current develop branch with all the changes into the PR, and re-tested the things. I've discovered a few more tiny bugs:

- [x] Editing a file externally messes up the affected main editor, which doesn't properly (re)load its theme
- [x] Editing a file externally twice messes up the loaded document in the document authority (Note: After the first editing, I ticked the "Always load changes" checkbox; have to verify this)
- [x] Renaming a file externally closes the file in the document tree but doesn't close a potentially now-empty pane, despite that it should (i.e. there is a pane with a single file open -> rename it externally -> Zettlr will close the file as expected -> Zettlr will, however, not close the corresponding pane that is now empty)

### CODE REVIEW

**Comment by nathanlesage** [comment]
Here we are; just merged a few final things and tested the PR out. In general, I feel that I've managed to make the code much less complex and achieved all the goals I set out to do.

I think this PR makes the entire app much more snappier, but since it's a somewhat bigger PR, I would love to have other people (including external people!) to do a code review on this beast before we merge.

I'll continue testing it out over the next days, but I have the feeling it's pretty stable already.

**Comment by nathanlesage** [comment]
Quick caveat: I'm experiencing weird app crashes on shutdown. Normally that indicates that I just forgot to close a chokidar instance (no danger, it's just annoying), but I couldn't find such a case yet … so if you find one, scream!

**Comment by kyaso** [comment]
Very interesting! I'm a bit busy these days but will sure make some room for this PR!

**Comment by nathanlesage** [comment]
Amazing, thank you!

And of course, I know that the PR is very large, so don't feel like you need to stress yourself too much! If we can get this out the door sometime before Christmas, that would be amazing :)

**Comment by kyaso** [comment]
Just a quick heads up, I'll finally have some time starting next week, so will be going over the PR then! (and sorry that I couldn't cram it in earlier)

**Comment by nathanlesage** [comment]
Wonderful, looking forward to your review!

**Comment by nathanlesage** [comment]
@kyaso First of all, I hope you had a few relaxing days so far. Do you already have an ETA for when you'll be able to review this PR? I have merged the new changes from develop into it yesterday, and it was a huge pain. I would love being able to merge this PR soon so that I can get the branches aligned again, and can focus on fixing a series of bugs! :)

Other than that, have a happy new year, and get safely to the other side!

**Comment by kyaso** [comment]
I'm really sorry I couldn't finish the review yet. I had to unexpectedly focus my attention on some private things (nothing serious!), so the PR was left dangling unfortunately.

I'll try to finish the review by this week then, so we can merge it ASAP!

Also, hope you had a smooth transition into 2024!

**Review by kyaso — COMMENTED** [review]
Whew, I finally managed to go over the PR! I left a few comments. Also, I have to admit that I couldn't fully grasp all (low-level) details yet, but I know that you know what you are doing ;) Do you plan to run some beta testing before the next release containing this PR?

File: source/app/service-providers/workspaces/index.ts [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+import generateStats, { type WorkspacesStatistics } from './generate-stats'
+
+export enum WORKSPACE_PROVIDER_EVENTS {
+  WorkspaceAdded = 'workspace-added',
+  WorkspaceChanged = 'workspace-changed',
+  WorkspaceRemoved = 'workspace-removed'
+}
+
+/**
+ * This class generates the Tray in the system notification area
```
[the remark this answers is not in the document]
Comment (nathanlesage) [reply source/app/service-providers/workspaces/index.ts:47]: Good point!

File: source/win-main/file-manager/file-tree.vue [hunk 2]
```diff
@@ -140,11 +142,13 @@ export default defineComponent({
...
   computed: {
+    ...mapStores(useWorkspacesStore),
     platform: function () {
       return process.platform
     },
-    fileTree: function (): MaybeRootDescriptor[] {
-      return this.$store.state.fileTree
+    fileTree: function (): AnyDescriptor[] {
+      return this.workspacesStore.roots.map(root => root.descriptor)
+      // return this.$store.state.fileTree
```
[the remark this answers is not in the document]
Comment (nathanlesage) [reply source/win-main/file-manager/file-tree.vue]: Couldn't find this commented code, but I think I've removed it properly!

**Comment by nathanlesage** [comment]
> Whew, I finally managed to go over the PR! I left a few comments. 

Amazing, thank you so much!

> Also, I have to admit that I couldn't fully grasp all (low-level) details yet, but I know that you know what you are doing ;)

… well, that's the hope :D I mean I did test it out extensively, so I hope that it's working well, but I haven't looked at the source for two weeks, so there's that. But since you couldn't find anything blatant, I don't think we're going to increase the amount of bugs any more that we already have.

> Do you plan to run some beta testing before the next release containing this PR?

Absolutely. This PR is large enough that it warrants release of a minor, i.e., it absolutely does need a few betas beforehand (and first a few weeks of nightlies).

I'm planning on releasing this PR alongside the preferences-UI overhaul as 3.1 in a few weeks.

***

Thanks for the code review; I've fixed the issues you mentioned and even managed to find and fix the bugs, so I'm going to merge so that this lock-file disaster on Git stops :D

**Pull request merged**